coderd/rbac/authz.go

```diff
@@ -137,6 +137,9 @@ func Filter[O Objecter](ctx context.Context, auth Authorizer, subject Subject, a
 			err := auth.Authorize(ctx, subject, action, o.RBACObject())
 			if err == nil {
 				filtered = append(filtered, o)
+			} else if ctx.Err() != nil {
+				// Exit early if the error comes from the context
+				return nil, ctx.Err()
```

Comment:
This code implicitly assumes that there are only 2 classes of errors that could be returned from `auth.Authorize`

1. the authorization completed, but the subject is not authorized
2. the context is Done

Prior to this change, this code implicitly assumed that there was only 1 class of errors.

It is difficult to be sure that the right answer is 2.  Could it be 3 classes?  4?

Honestly, I think a better design would be for `auth.Authorize` to return two values --- one that tells you whether the subject is authorized, and another to tell you whether or not there was an error checking the authorization.

Failing that, it's generally better for error checking to be explicit and specific:

```
err := auth.Authorize(...)
if xerrors.Is(err, rbac.Unauthorized) {
    continue
}
if err != nil {
    return nil, err
}
filtered = append(filtered, o)
```

Reply:
Actually, thinking about this more, it's not about implicit vs explicit.  It's about how the code reacts if `Authorize()` returns an _unexpected_ error.  By only handling the context case, we'd be the in the same boat again of swallowing the unexpected error and treating it like an unauthorized response.

If we instead handle the _expected_ errors, and return any others, we're less likely to incorrectly ignore them.

Reply:
Here are the 2 classes of error returned by `authorize`:

https://github.com/coder/coder/blob/6d41edceae37ac6e0494d3951944674b8f881dda/coderd/rbac/authz.go#L326-L333

Right now both are `ForbiddenWithInternal`. Maybe the correct solution here is to return something different on the `Eval` error.

Reply:
Then I can check for `rbac.IsUnauthorizedError()` on the `Filter` function loop

Reply:
Actually it gets a bit more annoying in the prepared one.

https://github.com/coder/coder/blob/6d41edceae37ac6e0494d3951944674b8f881dda/coderd/rbac/authz.go#L434-L434

:thinking: